lib/src/git.rs

```diff
@@ -987,73 +989,94 @@ fn update_git_head(
 
 /// Sets Git HEAD to the parent of the given working-copy commit and resets
 /// the Git index.
-pub fn reset_head(
-    mut_repo: &mut MutableRepo,
-    git_repo: &git2::Repository,
-    wc_commit: &Commit,
-) -> Result<(), git2::Error> {
+pub fn reset_head(mut_repo: &mut MutableRepo, wc_commit: &Commit) -> Result<(), GitExportError> {
+    let git_repo = get_git_repo(mut_repo.store()).ok_or(GitExportError::UnexpectedBackend)?;
+
     let first_parent_id = &wc_commit.parent_ids()[0];
     let first_parent = if first_parent_id != mut_repo.store().root_commit_id() {
         RefTarget::normal(first_parent_id.clone())
     } else {
         RefTarget::absent()
     };
-    if first_parent.is_present() {
-        let git_head = mut_repo.view().git_head();
-        let new_git_commit_id = Oid::from_bytes(first_parent_id.as_bytes()).unwrap();
-        let new_git_commit = git_repo.find_commit(new_git_commit_id)?;
-        if git_head != &first_parent {
-            git_repo.set_head_detached(new_git_commit_id)?;
-        }
 
-        let is_same_tree = if git_head == &first_parent {
-            true
-        } else if let Some(git_head_id) = git_head.as_normal() {
-            let git_head_oid = Oid::from_bytes(git_head_id.as_bytes()).unwrap();
-            let git_head_commit = git_repo.find_commit(git_head_oid)?;
-            new_git_commit.tree_id() == git_head_commit.tree_id()
-        } else {
-            false
-        };
-        let skip_reset = if is_same_tree {
-            // HEAD already points to a commit with the correct tree contents,
-            // so we only need to reset the Git index. We can skip the reset if
-            // the Git index is empty (i.e. `git add` was never used).
-            // In large repositories, this is around 2x faster if the Git index is empty
-            // (~0.89s to check the diff, vs. ~1.72s to reset), and around 8% slower if
-            // it isn't (~1.86s to check the diff AND reset).
-            let diff = git_repo.diff_tree_to_index(
-                Some(&new_git_commit.tree()?),
-                None,
-                Some(git2::DiffOptions::new().skip_binary_check(true)),
-            )?;
-            diff.deltas().len() == 0
-        } else {
-            false
-        };
-        if !skip_reset {
-            git_repo.reset(new_git_commit.as_object(), git2::ResetType::Mixed, None)?;
-        }
+    // If the first parent of the working copy has changed, reset the Git HEAD.
+    if mut_repo.git_head() != first_parent {
+        update_git_head(
+            &git_repo,
+            gix::refs::transaction::PreviousValue::Any,
```

Comment:
nit: can you leave a TODO comment about #3754? Ideally we should ensure that the existing HEAD ref points to the known object id (but we have to somehow deal with non-detached state.)

fwiw, maybe we can use `PreviousValue::MustExist` for a little bit of safety.